File: frontend/js/views/break_even.js
// Break-even / CVP analysis — how many units cover fixed costs, plus
// target-profit volume and margin of safety, via /calc/break-even.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';

const FIELDS = [
    ['fixed_costs_usd', 'Fixed costs ($/period)', 10000],
    ['price_per_unit_usd', 'Price per unit ($)', 50],
    ['variable_cost_per_unit_usd', 'Variable cost per unit ($)', 30],
    ['target_profit_usd', 'Target profit ($, optional)', 0],
    ['expected_units', 'Expected units sold (optional)', 800],
];

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 });
const units = (n) => Number(n).toLocaleString(undefined, { maximumFractionDigits: 1 });
const dash = (v, fn) => (v == null ? '—' : fn(v));

export async function renderBreakEven(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.breakeven.h1.title">// BREAK-EVEN ANALYSIS</span></h1>
        <p class="muted small" data-i18n="view.breakeven.hint.intro">
            The most basic small-business question: how many units must you sell to cover
            your costs? Each unit contributes price − variable cost toward the fixed costs
            (its contribution margin); break-even is where that accumulated contribution
            exactly covers them. Add a target profit to see the volume that earns it, and
            your expected sales to see the margin of safety. If the contribution margin is
            zero or negative, no volume ever breaks even — selling more only loses more.
        </p>
        <div class="chart-panel">
            <h2 data-i18n="view.breakeven.h2.inputs">Your numbers</h2>
            <form id="be-form" class="inline-form">
                ${FIELDS.map(([key, label, def]) => `
                    <label><span data-i18n="view.breakeven.label.${key}">${label}</span>
                        <input type="number" step="0.01" min="0" name="${key}" value="${def}" required></label>
                `).join('')}
                <button class="primary" type="submit" data-i18n="view.breakeven.btn.run">Calculate</button>
            </form>
        </div>
        <div id="be-result"></div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#be-form');
    form.addEventListener('submit', async (e) => {
        e.preventDefault();
        const fd = new FormData(e.target);
        const body = {};
        for (const [key] of FIELDS) body[key] = Number(fd.get(key)) || 0;
        try {
            const r = await api.calcBreakEven(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.breakeven.toast.error'), { level: 'error' });
        }
    });
    form.dispatchEvent(new Event('submit'));
}

function renderResult(mount, r) {
    const el = mount.querySelector('#be-result');
    const cmRatio = r.contribution_margin_ratio == null
        ? '—'
        : (Number(r.contribution_margin_ratio) * 100).toFixed(1) + '%';

    if (r.no_break_even) {
        el.innerHTML = `
            <div class="chart-panel">
                <h2 data-i18n="view.breakeven.h2.result">The result</h2>
                <div class="cards">
                    <div class="card neg"><div class="label" data-i18n="view.breakeven.card.cm">Contribution margin / unit</div>
                        <div class="value neg">${money(r.contribution_margin_usd)}</div></div>
                    <div class="card neg"><div class="label" data-i18n="view.breakeven.card.be_units">Break-even units</div>
                        <div class="value neg" data-i18n="view.breakeven.never">never</div></div>
                </div>
                <p class="muted small" data-i18n="view.breakeven.warn.no_be">Price is at or below variable cost — every unit loses money before fixed costs. Raise price or cut variable cost.</p>
            </div>
        `;
        applyUiI18n(el);
        return;
    }

    const mos = r.margin_of_safety_units;
    const mosCls = mos == null ? '' : mos >= 0 ? 'pos' : 'neg';
    const profit = r.profit_at_expected_usd;
    const profitCls = profit == null ? '' : profit >= 0 ? 'pos' : 'neg';
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.breakeven.h2.result">The result</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.breakeven.card.be_units">Break-even units</div>
                    <div class="value">${dash(r.break_even_units, units)}</div></div>
                <div class="card pos"><div class="label" data-i18n="view.breakeven.card.be_revenue">Break-even revenue</div>
                    <div class="value">${dash(r.break_even_revenue_usd, money)}</div></div>
                <div class="card"><div class="label" data-i18n="view.breakeven.card.cm">Contribution margin / unit</div>
                    <div class="value">${money(r.contribution_margin_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.breakeven.card.cm_ratio">Contribution margin ratio</div>
                    <div class="value">${cmRatio}</div></div>
                <div class="card"><div class="label" data-i18n="view.breakeven.card.target_units">Units for target profit</div>
                    <div class="value">${dash(r.units_for_target_profit, units)}</div></div>
                <div class="card"><div class="label" data-i18n="view.breakeven.card.target_revenue">Revenue for target profit</div>
                    <div class="value">${dash(r.revenue_for_target_profit_usd, money)}</div></div>
                <div class="card ${mosCls}"><div class="label" data-i18n="view.breakeven.card.mos_units">Margin of safety (units)</div>
                    <div class="value ${mosCls}">${dash(mos, units)}</div></div>
                <div class="card ${mosCls}"><div class="label" data-i18n="view.breakeven.card.mos_pct">Margin of safety (%)</div>
                    <div class="value ${mosCls}">${r.margin_of_safety_pct == null ? '—' : Number(r.margin_of_safety_pct).toFixed(1) + '%'}</div></div>
                <div class="card ${profitCls}"><div class="label" data-i18n="view.breakeven.card.profit">Profit at expected volume</div>
                    <div class="value ${profitCls}">${dash(profit, money)}</div></div>
            </div>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/break_even.rs
//! Break-even / cost-volume-profit (CVP) analysis.
//!
//! The most fundamental small-business question: how many units must I sell
//! to cover my costs? Each unit sold contributes `price − variable cost`
//! toward the fixed costs (the **contribution margin**); break-even is the
//! point where accumulated contribution exactly covers fixed costs.
//!
//!   * **Break-even units** = fixed costs / contribution margin per unit
//!   * **Break-even revenue** = fixed costs / contribution-margin ratio
//!   * **Target-profit units** = (fixed costs + target profit) / CM per unit
//!   * **Margin of safety** = how far expected sales sit above break-even
//!
//! When the contribution margin is zero or negative (each unit loses money
//! before fixed costs), there is no break-even at any volume — selling more
//! only deepens the loss. Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct BreakEvenInput {
    /// Total fixed costs for the period (rent, salaries, insurance, …).
    pub fixed_costs_usd: f64,
    pub price_per_unit_usd: f64,
    pub variable_cost_per_unit_usd: f64,
    /// Desired profit on top of covering fixed costs (default 0 = break-even).
    #[serde(default)]
    pub target_profit_usd: f64,
    /// Units you actually expect to sell, for the margin-of-safety figures
    /// (0 ⇒ skip those outputs).
    #[serde(default)]
    pub expected_units: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct BreakEvenResult {
    /// Price − variable cost: what each unit contributes to fixed costs.
    pub contribution_margin_usd: f64,
    /// Contribution margin as a fraction of price (0..1), or `None` if price ≤ 0.
    pub contribution_margin_ratio: Option<f64>,
    /// Units needed to break even; `None` when CM ≤ 0 (no break-even exists).
    pub break_even_units: Option<f64>,
    /// Revenue at the break-even point; `None` when CM ≤ 0.
    pub break_even_revenue_usd: Option<f64>,
    /// Units needed to reach the target profit; `None` when CM ≤ 0.
    pub units_for_target_profit: Option<f64>,
    /// Revenue needed to reach the target profit; `None` when CM ≤ 0.
    pub revenue_for_target_profit_usd: Option<f64>,
    /// Expected units − break-even units; `None` if no expected units / no CM.
    pub margin_of_safety_units: Option<f64>,
    /// Margin of safety as a percent of expected units; `None` likewise.
    pub margin_of_safety_pct: Option<f64>,
    /// Operating profit at the expected volume; `None` if no expected units.
    pub profit_at_expected_usd: Option<f64>,
    /// True when CM ≤ 0 — no volume ever breaks even.
    pub no_break_even: bool,
}

pub fn analyze(i: &BreakEvenInput) -> BreakEvenResult {
    let fixed = i.fixed_costs_usd.max(0.0);
    let price = i.price_per_unit_usd;
    let cm = price - i.variable_cost_per_unit_usd;
    let has_be = cm > 0.0;

    let cm_ratio = if price > 0.0 { Some(cm / price) } else { None };

    let be_units = if has_be { Some(fixed / cm) } else { None };
    let be_revenue = be_units.map(|u| u * price);

    let target_units = if has_be { Some((fixed + i.target_profit_usd) / cm) } else { None };
    let target_revenue = target_units.map(|u| u * price);

    let (mos_units, mos_pct, profit_at_expected) = if i.expected_units > 0.0 {
        let profit = i.expected_units * cm - fixed;
        let (mu, mp) = match be_units {
            Some(be) => {
                let mu = i.expected_units - be;
                let mp = if i.expected_units > 0.0 { mu / i.expected_units * 100.0 } else { 0.0 };
                (Some(mu), Some(mp))
            }
            None => (None, None),
        };
        (mu, mp, Some(profit))
    } else {
        (None, None, None)
    };

    BreakEvenResult {
        contribution_margin_usd: cm,
        contribution_margin_ratio: cm_ratio,
        break_even_units: be_units,
        break_even_revenue_usd: be_revenue,
        units_for_target_profit: target_units,
        revenue_for_target_profit_usd: target_revenue,
        margin_of_safety_units: mos_units,
        margin_of_safety_pct: mos_pct,
        profit_at_expected_usd: profit_at_expected,
        no_break_even: !has_be,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn base() -> BreakEvenInput {
        BreakEvenInput {
            fixed_costs_usd: 10_000.0,
            price_per_unit_usd: 50.0,
            variable_cost_per_unit_usd: 30.0,
            target_profit_usd: 0.0,
            expected_units: 0.0,
        }
    }

    #[test]
    fn basic_break_even_units_and_revenue() {
        // CM = 20; 10,000 / 20 = 500 units; 500 × $50 = $25,000 revenue.
        let r = analyze(&base());
        assert!((r.contribution_margin_usd - 20.0).abs() < 1e-9);
        assert!((r.break_even_units.unwrap() - 500.0).abs() < 1e-9);
        assert!((r.break_even_revenue_usd.unwrap() - 25_000.0).abs() < 1e-9);
        assert!(!r.no_break_even);
    }

    #[test]
    fn contribution_margin_ratio_and_revenue_consistency() {
        // CM ratio = 20/50 = 0.40; fixed / ratio = 10,000 / 0.40 = $25,000.
        let r = analyze(&base());
        assert!((r.contribution_margin_ratio.unwrap() - 0.40).abs() < 1e-9);
        let via_ratio = 10_000.0 / r.contribution_margin_ratio.unwrap();
        assert!((r.break_even_revenue_usd.unwrap() - via_ratio).abs() < 1e-6);
    }

    #[test]
    fn target_profit_adds_to_fixed_costs() {
        // (10,000 + 5,000) / 20 = 750 units.
        let r = analyze(&BreakEvenInput { target_profit_usd: 5_000.0, ..base() });
        assert!((r.units_for_target_profit.unwrap() - 750.0).abs() < 1e-9);
        assert!((r.revenue_for_target_profit_usd.unwrap() - 37_500.0).abs() < 1e-9);
    }

    #[test]
    fn margin_of_safety_against_expected_sales() {
        // Expected 800; BE 500 → 300 units of safety = 37.5%; profit = 800×20−10,000 = 6,000.
        let r = analyze(&BreakEvenInput { expected_units: 800.0, ..base() });
        assert!((r.margin_of_safety_units.unwrap() - 300.0).abs() < 1e-9);
        assert!((r.margin_of_safety_pct.unwrap() - 37.5).abs() < 1e-9);
        assert!((r.profit_at_expected_usd.unwrap() - 6_000.0).abs() < 1e-9);
    }

    #[test]
    fn negative_contribution_has_no_break_even() {
        // Price below variable cost: every unit loses money.
        let r = analyze(&BreakEvenInput {
            price_per_unit_usd: 30.0,
            variable_cost_per_unit_usd: 40.0,
            ..base()
        });
        assert!(r.no_break_even);
        assert!(r.break_even_units.is_none());
        assert!(r.units_for_target_profit.is_none());
        assert!((r.contribution_margin_usd - (-10.0)).abs() < 1e-9);
    }

    #[test]
    fn zero_fixed_costs_break_even_at_zero() {
        let r = analyze(&BreakEvenInput { fixed_costs_usd: 0.0, ..base() });
        assert!(r.break_even_units.unwrap().abs() < 1e-9);
        assert!(r.break_even_revenue_usd.unwrap().abs() < 1e-9);
    }

    #[test]
    fn expected_below_break_even_is_negative_safety_and_loss() {
        // Expected 400 < BE 500 → negative safety; profit = 400×20−10,000 = −2,000.
        let r = analyze(&BreakEvenInput { expected_units: 400.0, ..base() });
        assert!((r.margin_of_safety_units.unwrap() - (-100.0)).abs() < 1e-9);
        assert!(r.margin_of_safety_pct.unwrap() < 0.0);
        assert!((r.profit_at_expected_usd.unwrap() - (-2_000.0)).abs() < 1e-9);
    }
}
